vivarium/core/experiment.py

```diff
@@ -1423,7 +1426,7 @@ def run_derivers(self):
             if deriver:
                 # timestep shouldn't influence derivers
                 update, process_topology, state = self.process_update(
-                    path, deriver, 0)
+                    path, deriver, 0)  # TODO(jerry): Type Deriver doesn't have expected attribute 'schema'
```

Comment:
hmm, where do you see expected `'schema'`? They should have a `port_schema()` method that gives the schema. I see the same thing below for `process` -- is this called in `process_update`?

Reply:
That message is from PyCharm's type checker. In lieu of arg type declarations on the `process_update()` method, it looks into the method implementation and sees the expression `process.schema` and thus infers the `expected attribute 'schema'` requirement.

It's either catching a bug or something weird. The only code that assigns to something.schema is in `generate_paths()` which does `subprocess.schema = subprocess.get_schema()`. OK, that's weird, and breaking encapsulation. The `Process` class has no `schema` attribute but this reaches in and adds one, and the `process_update()` method assumes it's been added.

For better code structure, it could always call `subprocess.get_schema()` and if needed that method could cache the result for future quick access, or store the computed `schema` somewhere else, or at least `Process.__init__()` should initialize a `schema` attribute.

* I'll fix it in this PR if you like.

Reply:
Is there any reason to generate a process's schema dynamically like this? I don't think we expect to change processes after they're instantiated, so maybe a process should compute its schema only once in the constructor and then store that as an attribute. Schema computation involves a `copy.deepcopy`, so it'd be best to do it only once. Caching the result seems unnecessarily complicated though since it shouldn't change anyway